In [1]:
import numpy as np

k = 5
'''
problem1 function read the data, and format the data to numpy array

Return: the numpy array
'''
def data_reformation(filename):
    data = []
    ''''load data from .data file to list of list'''
    f=open(filename)
    sample=f.readlines()
    for i in range(len(sample)):
        sample[i] = sample[i].strip().split('\t')

    attributes = sample[0]
    sample.pop(0)
    #print("the sample we got right now is:", len(sample))
    for item in sample:
        list1=[]
        for i in range(len(item)):
            list1.append(np.float64(item[i]))
        list1.append(np.float64(1))
        data.append(list1)
    
    labels = []
    for instance in data:
        labels.append(instance[0])
        instance.pop(0)
    '''convert data to numpy array'''  
    prepared_data = np.array(data)
    labels = np.array(labels)
    
    return prepared_data, labels

training_data, training_data_labels = data_reformation("crime-train.txt")
print("shape of training_data:", training_data.shape, "and the shape of training_data_labels:", training_data_labels.shape,"the type of the training data is:", type(training_data))
testing_data, testing_data_labels = data_reformation("crime-test.txt")
print("shape of testing_data:", testing_data.shape, "and the shape of testing_data_labels:", testing_data_labels.shape)

shape of training_data: (1595, 96) and the shape of training_data_labels: (1595,) the type of the training data is: <class 'numpy.ndarray'>
shape of testing_data: (399, 96) and the shape of testing_data_labels: (399,)


In [2]:
training_data_folders = np.split(training_data, k);
training_data_labels_folders = np.split(training_data_labels, k)
lambda_values =[]
for i in range(10):
    if len(lambda_values) == 0:
        lambda_values.append(400)
    else:
        lambda_values.append(lambda_values[i-1]/2)


def cross_validation(lambda_value):
    ridge_validate_preds=[]
    rmses = []
    for i in range(len(training_data_folders)):
        training = list(training_data_folders)
        training.pop(i)
        training = np.concatenate(training, axis = 0)
        #print(training.shape)
        training_labels = list(training_data_labels_folders)
        training_labels.pop(i)
        #print(training_labels[0].shape)
        training_labels = np.concatenate(training_labels)
        #print(training_data_labels.shape)
        ridge_validate_preds.append(ridge_regression(training, training_labels,training_data_folders[i], lambda_value))
        rmses.append(show_rmse(ridge_validate_preds[i],training_data_labels_folders[i] ))
    
    return sum(rmses)/len(rmses) 

def ridge_regression(training,training_labels,validation,lambda_value):
    i_matrix = np.identity(96, dtype = np.float64())
    w =  np.linalg.inv((training).T.dot(training) + lambda_value * i_matrix).dot((training).T).dot(training_labels)
    return (validation).dot(w)

def show_rmse(pred, ans):
    assert len(pred) == len(ans)
    sqr_sum = 0
    for i, j in zip(pred, ans):
        sqr_sum = sqr_sum + (i-j)**2
    return (sqr_sum/len(pred))**(1/2)

def problem1():
    rmse_values = []
    for lambda_value in lambda_values:
        rmse_values.append(cross_validation(lambda_value))
        print("If lambda equals to:", lambda_value,"the average RMSE result I got is:",cross_validation(lambda_value))
    min_lambda = lambda_values[rmse_values.index(min(rmse_values))]
    print("\tthe smallest average RMSE result is", min(rmse_values),"and the lambda value is", min_lambda)
    ridge_test_optimal_lambda_pred = ridge_regression(training_data,training_data_labels, testing_data, min_lambda)
    ridge_train_optimal_lambda_pred = ridge_regression(training_data,training_data_labels, training_data, min_lambda)
    print("the training error of lambda =", min_lambda,"is", show_rmse(ridge_train_optimal_lambda_pred,training_data_labels))
    print("the testing error of lambda =", min_lambda,"is", show_rmse(ridge_test_optimal_lambda_pred,testing_data_labels))
problem1()

If lambda equals to: 400 the average RMSE result I got is: 0.14951277619075795
If lambda equals to: 200.0 the average RMSE result I got is: 0.14069401845317783
If lambda equals to: 100.0 the average RMSE result I got is: 0.13727677223186574
If lambda equals to: 50.0 the average RMSE result I got is: 0.13615594392755512
If lambda equals to: 25.0 the average RMSE result I got is: 0.13591585919851976
If lambda equals to: 12.5 the average RMSE result I got is: 0.13602237446230683
If lambda equals to: 6.25 the average RMSE result I got is: 0.13626786494076756
If lambda equals to: 3.125 the average RMSE result I got is: 0.136570956936428
If lambda equals to: 1.5625 the average RMSE result I got is: 0.13688764073197482
If lambda equals to: 0.78125 the average RMSE result I got is: 0.13717626806866834
	the smallest average RMSE result is 0.13591585919851976 and the lambda value is 25.0
the training error of lambda = 25.0 is 0.12879701459879783
the testing error of lambda = 25.0 is 0.1457465070

If lambda equals to: 400 the average RMSE result I got is: 0.14951277619075795
If lambda equals to: 200.0 the average RMSE result I got is: 0.14069401845317783
If lambda equals to: 100.0 the average RMSE result I got is: 0.13727677223186574
If lambda equals to: 50.0 the average RMSE result I got is: 0.13615594392755512
If lambda equals to: 25.0 the average RMSE result I got is: 0.13591585919851976
If lambda equals to: 12.5 the average RMSE result I got is: 0.13602237446230683
If lambda equals to: 6.25 the average RMSE result I got is: 0.13626786494076756
If lambda equals to: 3.125 the average RMSE result I got is: 0.136570956936428
If lambda equals to: 1.5625 the average RMSE result I got is: 0.13688764073197482
If lambda equals to: 0.78125 the average RMSE result I got is: 0.13717626806866834
	the smallest average RMSE result is 0.13591585919851976 and the lambda value is 25.0
the training error of lambda = 25.0 is 0.12879701459879783
the testing error of lambda = 25.0 is 0.14574650707057987

In [3]:
from random import *
epsilon = 10**-5
#linear test error: 0.14550044963059267 
#linear train error: 0.12928590193840733 
#alpha = 0.00000845
def problem2(samples):
    alpha = 0.00000845
    w_initial = []
    for i in range(96):
        w_initial.append(np.float64(random()))
    w_t = np.array(w_initial)
    print("w_t:",w_t,"\n")
    w_t1 = w_t + alpha*((training_data).T.dot(training_data_labels-((training_data).dot(w_t))))
    print("w_t1",w_t1.shape,"\n")
    while weight_compare(w_t, w_t1) == False:
        w_t = w_t1
        #print("w_t:",w_t,"\n")
        w_t1 = w_t + alpha*((training_data).T.dot(training_data_labels-((training_data).dot(w_t))))
        #print("w_t1:",w_t1,"\n")
    return (samples).dot(w_t1)

def weight_compare(w1, w2):
    assert len(w1) == len(w2)
    for i in range(len(w1)):
        if abs(w1[i]-w2[i])>=(epsilon):
            return False
    return True

def problem3(samples):
    alpha = 0.0000085
    w_initial = []
    for i in range(96):
        w_initial.append(np.float64(random()))
    w_t = np.array(w_initial)
    rmse_values = []
    for lambda_value in lambda_values:
        rmse_values.append(cross_validation(lambda_value))
        #print("If lambda equals to:", lambda_value,"the average RMSE result I got is:",cross_validation(lambda_value))
    min_lambda = lambda_values[rmse_values.index(min(rmse_values))]
    i_matrix = np.identity(96, dtype = np.float64())
    w_t1 = w_t + alpha*((training_data).T.dot(training_data_labels-((training_data).dot(w_t))) - min_lambda*w_t)
    while weight_compare(w_t, w_t1) == False:
        w_t = w_t1
        #print("w_t:",w_t,"\n")
        w_t1 = w_t + alpha*((training_data).T.dot(training_data_labels-((training_data).dot(w_t))) - min_lambda*w_t)
        #print("w_t1:",w_t1,"\n")
    return (samples).dot(w_t1)
linear_train_pred = problem2(training_data)
linear_test_pred = problem2(testing_data)
ridge_test_pred = problem3(testing_data)
ridge_train_pred = problem3(training_data)
print("linear test error:",show_rmse(linear_test_pred, testing_data_labels),"\n")
print("linear train error:",show_rmse(linear_train_pred, training_data_labels),"\n")
print("ridge train error:",show_rmse(ridge_train_pred, training_data_labels),"\n")
print("ridge test error:",show_rmse(ridge_test_pred, testing_data_labels),"\n")

w_t: [0.01051792 0.55135813 0.52269449 0.5661685  0.25271851 0.79230716
 0.38895629 0.60838067 0.63140261 0.02892042 0.68455986 0.85470498
 0.08975153 0.5868919  0.88927338 0.94680339 0.43563193 0.86242412
 0.09258226 0.5931769  0.85601854 0.83310365 0.07571838 0.95690498
 0.84478696 0.95936511 0.05882305 0.75092406 0.87466923 0.20121572
 0.89509245 0.85738438 0.57261735 0.75825825 0.89668246 0.26093171
 0.37629578 0.16761548 0.81550182 0.27243789 0.31501051 0.38238638
 0.93965645 0.89854976 0.43339247 0.07027847 0.44061247 0.08993339
 0.14014423 0.62500647 0.6609046  0.23664367 0.4522918  0.94631494
 0.41319386 0.74297895 0.30129232 0.18549921 0.97460458 0.06305675
 0.69342925 0.45586226 0.36306739 0.40786626 0.45763638 0.16604318
 0.917921   0.23792413 0.38806221 0.5385568  0.77583291 0.40111849
 0.85453367 0.71735098 0.11281745 0.97044276 0.6345596  0.77885326
 0.53031864 0.92605493 0.41286201 0.71515847 0.60760953 0.34434347
 0.4209883  0.24808751 0.22000989 0.70749169 0.28517014 0

problem 2: 
alpha = 0.00000845
linear test error: 0.14550044963059267 
linear train error: 0.12928590193840733 
In the lab4 I get linear test error: 0.1458346449094907 and linear train error: 0.12768967421762192. It is really close

problem3:
alpha = 0.0000085
The closes result I get is:
ridge train error: 0.12887069926895178 
ridge test error: 0.14570432563991323 
It is very close to the result in Exercise 1
